src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpRetryHandler.cs

```diff
@@ -66,25 +67,43 @@ public class HttpRetryHandler : IHttpRetryHandler
                         // disposing it.
                         response?.Dispose();
 
-                        var timeoutMessage = string.Format(
-                            CultureInfo.CurrentCulture,
-                            Strings.Http_Timeout,
-                            requestMessage.Method,
-                            requestUri,
-                            (int)request.RequestTimeout.TotalMilliseconds);
-
                         log.LogInformation("  " + string.Format(
                             CultureInfo.InvariantCulture,
                             Strings.Http_RequestLog,
                             requestMessage.Method,
                             requestUri));
 
+                        // Issue the request.
+                        var timeoutMessage = string.Format(
+                            CultureInfo.CurrentCulture,
+                            Strings.Http_Timeout,
+                            requestMessage.Method,
+                            requestUri,
+                            (int)request.RequestTimeout.TotalMilliseconds);
                         response = await TimeoutUtility.StartWithTimeout(
                             timeoutToken => request.HttpClient.SendAsync(requestMessage, request.CompletionOption, timeoutToken),
                             request.RequestTimeout,
                             timeoutMessage,
                             cancellationToken);
 
+                        // Wrap the response stream so that the download can timeout and the HTTP
+                        // semaphore is respected.
+                        Stream networkStream;
+                        if (response.Content != null)
+                        {
+                            networkStream = await response.Content.ReadAsStreamAsync();
+                        }
+                        else
+                        {
+                            networkStream = Stream.Null;
```

Comment:
This is actually not necessary. A null `.Content` property can be left as-is (this often happens in tests that don't completely fill out the `HttpResponseMessage`.